### PR TITLE
Check Last-Modified before update tenant index

### PR DESCRIPTION
**Pull request by vporoshok**
#### What this PR does

Update (download, unzip and parse) tenant index only if file has changed in storage (changed Last-Modified time). It is useful in case if we need low-latency block detection. With this changes we may use very short update intervals (ex. 1 second).

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pracucci** [comment]
👋  Hi! Thanks for your interest in the project.

Could you share more details about your use case? I have some thoughts in random order, but I fundamentally don't understand the use case, so I would like to ear more about it first.

Random thoughts:
- Mimir has been designed to not require an up-to-date view over the bucket. What's your use case where you need it?
- The bucket index is updated by the compactor. Even if you reduce the blocks cleaner interval to 1s, it will realistically some time to update the bucket index. What's your expected max latency between when something change in the bucket and when queriers and store-gateways detect it?
- This PR introduces a S3 call to check the last modified, before actually reading the index. An S3 API call to fetch attributes (HEAD) is billed as a GET. Intra region bandwidth between S3 and compute machines is not billed, so costing wise there should be no difference (pretty much the same for other cloud providers). I assume the check on last modified has been introduced to reduce CPU utilization, but what's the actual impact? How many tenants do you have in your cluster?

**Comment by pracucci** [comment]
Thanks for providing more details.

> We have our own agent, and it can deliver metrics after long storage unavailability. It's uploads block into S3 directly

If the agent deliver metrics "after long storage unavailability" is that critical to have these metrics available for querying 1 second after the block has been uploaded?

> Because of this, for us, important low sync interval.

Could you share some numbers to let me further understand your use case, please?

- How low the sync interval is expected?
- What value you set for `-compactor.cleanup-interval`?
- If you build this PR and deploy it to a querier or store-gateway, what's the drop in CPU utilization?

_Thanks!_

**Comment by vporoshok** [comment]
> * What value you set for -compactor.cleanup-interval?

1 minute. So

> * How low the sync interval is expected?

We expect sub 2 minutes data availability (we use 1 minute point intervals).

> * If you build this PR and deploy it to a querier or store-gateway, what's the drop in CPU utilization?

Will provide numbers in closest time

**Comment by vporoshok** [comment]
We found that storage gateway doesn't use `bucketindex.Loader`, so for now this changes doesn't improve performance in our case. We need more research. I'll back with new information soon.

**Comment by vporoshok** [comment]
We've got 2x decrease CPU usage after using index loader (with last-modified check) in store-gateway:
image (1)

File: pkg/storage/tsdb/bucketindex/storage.go [hunk 1]
```diff
@@ -20,14 +21,76 @@ import (
...
+	if !attrs.LastModified.After(lastModified) {
+		return nil, ErrIndexNotModified
+	}
+	idx, err := readIndex(ctx, userBkt, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &IndexWithLastModified{
+		Index:        idx,
+		LastModified: attrs.LastModified,
```
Comment (andyasp) [inline pkg/storage/tsdb/bucketindex/storage.go:55]: Just wanted to add a note about something I noticed while giving this a quick glance: this is mixing the last modified time from the HEAD and the result of the GET, which is potentially confusing and could cause extra reads when they're not consistent. Looks like this is a limitation of the `BucketReader` interface.

Comment (andyasp) [inline pkg/storage/tsdb/bucketindex/storage.go:55]: Just wanted to add a note about something I noticed while giving this a quick glance: this is mixing the last modified time from the HEAD and the result of the GET, which is potentially confusing and could cause extra reads when they're not consistent. Looks like this is a limitation of the `BucketReader` interface.

Comment (andyasp) [reply pkg/storage/tsdb/bucketindex/storage.go:55]: Ideally this would use a single GET with an `If-Modified-Since` header.

**Comment by pstibrany** [comment]
To make all checks pass again, please rebase this PR on top of `main`. Thank you.

**Comment by johannaratliff** [comment]
Before merging, can you rebase against main and pop the changelog entry into the ## main / unreleased so it doesn't include in v2.1.0? thank you

**Comment by pracucci** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

**Pull request closed without merging**